### ISSUE TITLE
Vectorized Differentiation for Gradient and Jacobian calculations

### ISSUE THREAD

**Issue by joshuaeh** [pnnl/neuromancer#173]
As of `torch==2.0`, the `torch.func` module contains functions to:
1. Parametrize `torch.nn.Module` instances as a function of inputs and parameter values
2. Vectorize jacobian and multi-output gradient calculations

The current `gradients.py` module calculates the jacobian as follows:
https://github.com/pnnl/neuromancer/blob/4f8bece3c812fd345fa52dcb4d817ef5acee4e35/src/neuromancer/gradients.py#L25-L38

This vectorization has increased my calculations in other projects significantly. An example implementation would be

```python
from torch.func import functional_call, jacrev

def jacobian(f, x)->dict:
    return jacrev(f)(x)
```  

Where `f` is a `torch.nn.Module` or any other function with tensor-valued inputs and outputs. `jacrev` would work inplace for calculating the Jacobian w.r.t. input features. To calculate the Jacobian w.r.t. `torch.nn.Module` parameters, the functional version of the model can be created using `functional_call()`

```python
f = lambda parameters, inputs: functional_call(model, parameters, inputs)
J = jacobian(f, (parameters, inputs))
```  

A minimum working example:

```python
import torch
from torch.func import functional_call, jacrev

def jacobian(f, x)->dict:
    return jacrev(f)(x)

model = torch.nn.Linear(3, 3)
x = torch.randn(3)

# calculate jacobian w.r.t. input data
j = jacobian(model, x)
print("Jacobian w.r.t. input data")
print(f"{dict(model.named_parameters())=}")
print(f"{j=}")

# calculate jacobian w.r.t. model parameters

def f(params_inputs):
    """returns functional form of the model"""
    return functional_call(model, *params_inputs)

j = jacobian(f, (dict(model.named_parameters()), x))
print("\nJacobian w.r.t. model parameters")
print(f"{dict(model.named_parameters())=}")
print(f"{j=}")
```

I can create a pull request, test conflicts, add examples etc. if it would be a meaningful addition. I'd be happy to hear thoughts on why this wouldn't be a good idea.

**Comment by drgona** [comment]
Dear @joshuaeh thank you very much for this suggestion. This contribution would be of high interest to the library, and your effort in creating new PR would be greatly appreciated!

**Comment by RBirmiwal** [comment]
Closing this issue. @joshuaeh feel free to still create that PR
